# 錯誤更正碼 1004

## Finite Fields

Linear codes 要有定義，必須先有 linear algebra  
要有 linear algebra，必須要有 fields  
今天介紹 **有限域（finite field，也稱有限體）**，就是 coding 版本的實數跟複數

食譜
- 從模質數的運算可以定義 $\mathbb F_p$。
- 線性代數說元素個數一定是 $p^m$。
- 用不可約多項式建構 $\mathbb F_4$、$\mathbb F_8$ 與 $\mathbb F_9$ 等等。
- 用有限域計算 codewords 與 syndromes。

### 想要 alphabet 有代數結構

對一般 block code，alphabet $\mathcal A$ 只需要是一個集合。  
Hamming distance 只比較兩個 symbols 是否相同，不需要加法或乘法。

但到了 linear code，我們要
$$c = uG,\qquad s = yH,\qquad y = c + e.$$
這些公式需要 symbols 之間可以相加，也需要用 symbols 當作純量相乘。  
有限域提供這些運算，使線性代數進到房間裡。

### Field 的定義

一個 **域（field）** 是一個集合 $F$，它有加法與乘法，並且滿足以下規則。  

1. 加法與乘法都有交換律與結合律。
2. $a + 0 = a$，$a \cdot 1 = a$。
3. 每個 $a$ 都有加法反元素 $-a$，滿足 $a + (-a) = 0$。
4. 每個 $a \ne 0$ 都有乘法反元素 $a^{-1}$，滿足 $aa^{-1} = 1$。
5. 乘法對加法有分配律：$a(b + c) = ab + ac$。

因此，只要 $b \ne 0$，就能定義除法
$$\frac ab = ab^{-1}.$$
域中有消去律：若 $a \ne 0$ 且 $ab = ac$，則 $b = c$。

特別地，域中沒有非零元素相乘得到零的情形：
$$ab = 0\quad\Longrightarrow\quad a = 0\text{ 或 }b = 0.$$
證明：當 $a \ne 0$ 時，將等式兩側乘上 $a^{-1}$，便得到 $b = 0$。  
若集合 $F$ 的元素個數有限，就稱為有限域。

## Prime Fields

### 模一個質數

給定質數 $p$，以 $0, 1, \ldots, p - 1$ 代表模 $p$ 的各個餘數類。  
先做整數的加法或乘法，再取模 $p$，得到
$$\mathbb F_p = \mathbb Z/p\mathbb Z.$$
例如在 $\mathbb F_5$ 中，
$$3 + 4 = 2,\qquad 3 \cdot 4 = 2,\qquad -2 = 3.$$
這些等式的左右兩側都表示域中的元素。  
當需要寫整數之間的關係時，使用 $3 + 4 \equiv 2 \pmod 5$。

### 為什麼每個非零元素都能除？

若 $1 \le a < p$，因為 $p$ 是質數，所以 $\gcd(a, p) = 1$。  
整數的 Bézout identity 給出整數 $u, v$，使
$$ua + vp = 1.$$
取模 $p$ 後，得到 $ua = 1$，所以 $u$ 的餘數類就是 $a^{-1}$。

例如 $2 \cdot 3 = 1$ 在 $\mathbb F_5$ 中成立，因此
$$2^{-1} = 3,\qquad 3^{-1} = 2,\qquad \frac32 = 3 \cdot 3 = 4.$$

$\mathbb F_5$ 的非零元素與反元素如下：

| $a$ | $1$ | $2$ | $3$ | $4$ |
|---|---:|---:|---:|---:|
| $a^{-1}$ | $1$ | $3$ | $2$ | $4$ |

在 $\mathbb F_2 = \{0, 1\}$ 中，$1 + 1 = 0$，加法就是 XOR。  
唯一的非零元素是 $1$，且 $1^{-1} = 1$。  
前面的 binary linear codes 都使用這個域。

### 模一個合數

模 $n$ 的餘數類也能相加、相乘，形成環 $\mathbb Z/n\mathbb Z$。  
其中 $a$ 有乘法反元素的條件是 $\gcd(a, n) = 1$。

例如在 $\mathbb Z/4\mathbb Z$ 中，
$$2 \cdot 2 = 0.$$
元素 $2$ 沒有乘法反元素，因此這個環不滿足域的定義。  
要建構四元素的域，我們會使用另一套加法與乘法。

## Characteristic 與 Prime Powers

### 把 $1$ 一直相加

對任意的有限域 $F$，考慮
$$0,\quad 1,\quad 1 + 1,\quad 1 + 1 + 1,\quad \ldots.$$
因為元素個數有限，必定有兩個和相同；相減後，就得到某個正整數次的 $1$ 相加為零。  
使下式成立的最小正整數稱為 $F$ 的 **characteristic（特徵）**：
$$\underbrace{1 + \cdots + 1}_{p\text{ 個}} = 0.$$
例如 $\mathbb F_2$ 的 characteristic 是 $2$，$\mathbb F_5$ 的 characteristic 是 $5$。

### Characteristic 一定是質數

假設最小正整數 $p$ 可以分解成 $p = ab$，其中 $1 < a, b < p$。  
域中的兩個元素 $a \cdot 1$ 與 $b \cdot 1$ 都非零，因為相加次數還沒到 $p$。  
但分配律給出
$$(a \cdot 1)(b \cdot 1) = (ab) \cdot 1 = p \cdot 1 = 0.$$
這與域中沒有非零元素相乘得到零的性質矛盾，所以 $p$ 必定是質數。

由 $0, 1, \ldots, (p - 1) \cdot 1$ 組成的集合，就是 $F$ 中的一份 $\mathbb F_p$。  
它稱為 **prime subfield**，是 $F$ 最小的子域。

### 有限域是 prime subfield 上的向量空間

令 $F$ 的 characteristic 為 $p$。  
用 $F$ 的加法與 $\mathbb F_p$ 中的純量乘法，就能把 $F$ 看成 $\mathbb F_p$ 上的向量空間。

若一組基底是 $b_1, \ldots, b_m$，每個元素都有唯一的表示
$$a_1b_1 + \cdots + a_mb_m,\qquad a_i \in \mathbb F_p.$$
每個係數有 $p$ 種選擇，因此
$$\boxed{|F| = p^m.}$$
有限域的元素個數一定是質數冪，且 $m \ge 1$。

例如 $4 = 2^2$、$8 = 2^3$、$9 = 3^2$ 都可能是有限域的大小。  
$6$、$10$、$12$ 都不是質數冪，所以不存在這些大小的有限域。

**有限域的存在與唯一性定理：**  
對每個質數 $p$ 與正整數 $m$，都存在含有 $p^m$ 個元素的域；任意兩個這樣的域都同構。  
也就是有一個一一對應，同時保留加法、乘法、$0$ 與 $1$。

因此可以把它記為 $\mathbb F_{p^m}$，或 $\operatorname{GF}(p^m)$。  
接下來先介紹多項式的商域，再證明這個存在與唯一性定理。

## 用多項式建構有限域

### 多項式的除法與餘數

$\mathbb F_p[x]$ 表示係數在 $\mathbb F_p$ 中的多項式。  
多項式的係數相加、相乘時，也使用 $\mathbb F_p$ 的運算。

固定一個 monic polynomial $f(x)$，即最高次項係數為 $1$，令 $\deg f = m$。  
多項式除法給出唯一的 $q(x)$ 與 $r(x)$，使
$$g(x) = q(x)f(x) + r(x),\qquad r = 0\text{ 或 }\deg r < m.$$
我們用這個餘數代表 $g(x)$，稱為取模 $f(x)$。

兩個多項式代表同一個元素，當且僅當它們的差是 $f(x)$ 的倍數。  
所有這些餘數類組成 **quotient ring**
$$\mathbb F_p[x]/(f(x)).$$
加法與乘法的規則是先做多項式運算，再取模 $f(x)$。  
每個元素都有唯一的代表
$$a_0 + a_1x + \cdots + a_{m - 1}x^{m - 1},\qquad a_i \in \mathbb F_p.$$
因此這個 quotient ring 有 $p^m$ 個元素。  
把代表換成 $g(x) + h(x)f(x)$，不會改變運算後的餘數，所以運算是明確的。

### Irreducible polynomial

若次數至少為 $1$ 的多項式 $f(x)$ 不能分解成兩個正次數多項式的乘積，  
就稱它在 $\mathbb F_p$ 上 **irreducible（不可約）**。

例如 $x^2 + x + 1 \in \mathbb F_2[x]$ 在 $0$ 與 $1$ 都不為零，所以沒有一次因子。  
二次多項式若可約，就一定有一次因子，因此這個多項式不可約。

同樣的判斷也適用於三次多項式：二次或三次多項式不可約，當且僅當它在係數域中沒有根。  
四次以上的多項式還可能分解成兩個次數至少為 $2$ 的因子，判斷時要檢查這些可能性。

### 不可約性讓每個非零元素都有反元素

假設 $f$ 不可約，且 $g$ 是非零的代表，$\deg g < \deg f$。  
此時 $\gcd(g, f) = 1$；多項式的 extended Euclidean algorithm 給出
$$u(x)g(x) + v(x)f(x) = 1.$$
取模 $f(x)$，得到 $u(x)g(x) = 1$，所以 $u(x)$ 的餘數類就是 $g(x)^{-1}$。

因此
$$\boxed{\mathbb F_p[x]/(f(x))\text{ 是含有 }p^m\text{ 個元素的域，}\quad \deg f = m.}$$
這個建構所需的條件是 $f$ 在 $\mathbb F_p$ 上不可約。

若 $f$ 可約，寫成 $f(x) = a(x)b(x)$，其中 $0 < \deg a, \deg b < \deg f$。  
在取模 $f$ 後，$a$ 與 $b$ 都代表非零元素，卻滿足 $ab = 0$。

例如在 $\mathbb F_2[x]/(x^2 + 1)$ 中，
$$(x + 1)^2 = x^2 + 1 = 0.$$
因為 $x^2 + 1 = (x + 1)^2$ 在 $\mathbb F_2[x]$ 中可約，所以這個 quotient ring 有四個元素，卻不是域。

### 多項式的 root bound

**定理：** 域上的非零多項式 $f(x)$ 若次數為 $d$，在這個域中至多有 $d$ 個不同的根。

若 $f(a) = 0$，多項式除法給出 $f(x) = (x - a)g(x)$。  
對另一個根 $b \ne a$，有
$$0 = f(b) = (b - a)g(b).$$
因為 $b - a \ne 0$ 且可以消去，所以 $g(b) = 0$。  
每取出一個不同的根，就少一次；重複這個過程，便得到 root bound。  
域中非零元素可消去的性質，是這個論證的關鍵。

## 有限域的存在與唯一性

### 在 extension field 中加入根

若域 $K$ 包含域 $E$，且在 $E$ 上使用相同的運算，就稱 $K$ 是 $E$ 的 **extension field（擴張域）**。  
前面商域的 Bézout 證明適用於任意域 $E$：若 $g(t) \in E[t]$ 是不可約多項式，則
$$E[t]/(g(t))$$
是一個域，並透過常數多項式包含一份 $E$。

把 $t$ 的餘數類記為 $\alpha$，就得到 $g(\alpha) = 0$。  
因此，取商域可以把不可約多項式的一個根加入原來的域。

任意正次數多項式都有不可約因子：若它可約，就繼續分解，因子的次數逐次降低。  
給定 $h(x) \in E[x]$，取它的一個不可約因子，利用商域取得根 $\alpha$，再除去因子 $x - \alpha$。  
對剩下的多項式重複這個程序。

每次除去一個一次因子，剩餘次數就減少 $1$，所以有限次之後一定能完全分解。  
得到的擴張域 $K$ 由加入的根生成，稱為 $h$ 在 $E$ 上的 **splitting field（分裂域）**。  
它是包含 $E$ 與 $h$ 的全部根的最小子域。  
若 $E$ 有限，每次加入根仍是有限維的商域，所以 $K$ 也有限。

### 存在性：$x^{p^m} - x$ 的根形成域

固定質數 $p$ 與正整數 $m$，令 $q = p^m$，考慮
$$h(x) = x^q - x \in \mathbb F_p[x].$$
利用剛才的程序，取得它的一個分裂域 $K$。  
在 $K[x]$ 中，$h$ 可以分解成 $q$ 個一次因子的乘積。

接下來證明這些根都不同，且全部根在 $K$ 的運算下形成一個域。

對域上的多項式，定義 **formal derivative（形式導數）**：
$$\left(\sum_{j = 0}^d c_j x^j\right)'
= \sum_{j = 1}^d j c_j x^{j - 1}.$$
由展開可以驗證乘法法則 $(uv)' = u'v + uv'$。  
若 $a$ 是重根，則 $h(x) = (x - a)^2v(x)$；對這個分解微分，就得到 $h'(a) = 0$。

因為 $K$ 的 characteristic 是 $p$，而 $p$ 整除 $q$，
$$h'(x) = qx^{q - 1} - 1 = -1.$$
所以 $h$ 沒有重根；它在 $K$ 中恰好有 $q$ 個不同的根。

在 characteristic 為 $p$ 的域中，$p$ 整除 $\binom pj$，其中 $0 < j < p$。  
二項式定理因此給出
$$(a + b)^p = a^p + b^p.$$
連續取 $m$ 次 $p$ 次方，便得到
$$(a + b)^q = a^q + b^q.$$
這個等式也保留加法反元素，所以 $(a - b)^q = a^q - b^q$。  
乘法則直接滿足 $(ab)^q = a^q b^q$。

令 $S$ 是 $h$ 在 $K$ 中的全部根：
$$S = \{a \in K : a^q = a\},\qquad |S| = q.$$
首先 $0, 1 \in S$。若 $a, b \in S$，則
$$(a - b)^q = a^q - b^q = a - b,$$
$$(ab)^q = a^q b^q = ab.$$
因此 $S$ 對減法與乘法封閉。  
對減法封閉也給出 $-a = 0 - a \in S$，以及 $a + b = a - (-b) \in S$。

若 $a \in S$ 且 $a \ne 0$，它在 $K$ 中的反元素滿足
$$(a^{-1})^q = (a^q)^{-1} = a^{-1}.$$
所以 $a^{-1}$ 也在 $S$ 中。

結合前面的封閉性，$S$ 繼承 $K$ 的域運算，且恰好有 $q = p^m$ 個元素。  
這就證明了每個質數冪大小的有限域都存在。

### 每個 $q$ 元素域都包含 $x^q - x$ 的全部根

反過來，令 $F$ 是任意一個有 $q$ 個元素的域。  
固定 $a \ne 0$，乘上 $a$ 會重新排列 $F$ 的全部非零元素。  
令 $P$ 是所有非零元素的乘積，則 $P \ne 0$，而
$$P = \prod_{u \in F^\times}(au)
= a^{q - 1}\prod_{u \in F^\times}u
= a^{q - 1}P.$$
消去 $P$，得到 $a^{q - 1} = 1$。  
因此 $a^q = a$ 對每個 $a \in F$ 都成立，包含 $a = 0$。

所以 $F$ 的 $q$ 個元素都是 $x^q - x$ 的根，且
$$x^q - x = \prod_{a \in F}(x - a).$$
兩側都是 monic、次數為 $q$ 的多項式，右側列出左側的全部根。

因此，任意 $q$ 元素域都是 $x^q - x$ 在其 prime subfield 上的一個分裂域。  
為了證明這些域同構，我們接下來說明如何把根的對應延伸成域的同構。

### Minimal polynomial 與加入根後的域

若 $\alpha$ 在某個擴張域中，且某個非零的 $E$ 係數多項式在 $\alpha$ 取值為零，  
就稱 $\alpha$ 在 $E$ 上是 **algebraic（代數的）**。

選擇滿足 $f(\alpha) = 0$、次數最小的 monic polynomial $f \in E[x]$，  
稱為 $\alpha$ 在 $E$ 上的 **minimal polynomial（最小多項式）**。  
它必定不可約：若 $f = uv$ 且兩個因子次數都較小，  
由 $u(\alpha)v(\alpha) = 0$ 可知其中一個在 $\alpha$ 為零，與最小性矛盾。

將任意 $g \in E[x]$ 除以 $f$，寫成
$$g = vf + r,\qquad r = 0\text{ 或 }\deg r < \deg f.$$
若 $g(\alpha) = 0$，則 $r(\alpha) = 0$；由 $f$ 的最小性，必有 $r = 0$。  
因此
$$g(\alpha) = 0\quad\Longleftrightarrow\quad f \mid g.$$
這也保證 minimal polynomial 唯一：兩個 monic 最小多項式必定互相整除。

令 $E(\alpha)$ 表示包含 $E$ 與 $\alpha$ 的最小子域。  
代入 $\alpha$ 給出對應
$$E[x]/(f(x)) \longrightarrow E(\alpha),\qquad [g(x)] \longmapsto g(\alpha).$$
剛才的整除條件保證這個對應是明確且一一的，並保留加法與乘法。  
左側是域，所以其像也是域；它包含 $E$ 與 $\alpha$，因此像恰好是 $E(\alpha)$。

所以
$$\boxed{E(\alpha) \cong E[x]/(f(x)).}$$

### 同構可以沿著根的對應延伸

假設已有域同構 $\phi : E \longrightarrow E'$，且 $\alpha$ 在 $E$ 上的 minimal polynomial 是 $f$。  
把 $f$ 的每個係數套用 $\phi$，得到 $\widetilde f \in E'[x]$。  
它在 $E'$ 上不可約，因為任何分解都能經 $\phi^{-1}$ 還原成 $f$ 的分解。

若 $\beta$ 是 $\widetilde f$ 在某個擴張域中的根，它在 $E'$ 上的 minimal polynomial 就是 $\widetilde f$。  
對每個 $g \in E[x]$，同樣把係數套用 $\phi$，記結果為 $\widetilde g$。

商域的表示給出域同構
$$\Phi : E(\alpha) \longrightarrow E'(\beta),\qquad
g(\alpha) \longmapsto \widetilde g(\beta).$$
若兩個 $g$ 在 $\alpha$ 的值相同，它們的差被 $f$ 整除；  
係數映射後的差就被 $\widetilde f$ 整除，因此在 $\beta$ 的值也相同。  
反方向用 $\phi^{-1}$ 做相同的構造，所以 $\Phi$ 有逆映射。

這個同構滿足 $\Phi|_E = \phi$，且 $\Phi(\alpha) = \beta$。  
也就是既有的同構可以延伸到加入根後的域。

### 唯一性：逐一把根對應起來

令 $F$ 與 $F'$ 都有 $q = p^m$ 個元素。  
由前面的質數冪論證，兩者的 characteristic 都是 $p$。  
先把兩者的 prime subfields 按 $a \cdot 1 \mapsto a \cdot 1$ 對應，記為 $\phi_0$。

列出 $F$ 的全部元素 $\alpha_1, \ldots, \alpha_q$，並依序加入：
$$E_0 = \mathbb F_p,\qquad E_i = E_{i - 1}(\alpha_i),\qquad E_q = F.$$
我們逐步把 $\phi_0$ 延伸成 $E_i$ 到 $F'$ 的某個子域的同構。

假設已經有 $\phi_{i - 1} : E_{i - 1} \longrightarrow E'_{i - 1} \subseteq F'$。  
令 $f_i$ 是 $\alpha_i$ 在 $E_{i - 1}$ 上的 minimal polynomial。  
因為 $\alpha_i^q = \alpha_i$，前面的整除條件給出
$$f_i(x) \mid x^q - x.$$
把係數經 $\phi_{i - 1}$ 映射，得到 $\widetilde f_i$。  
這個同構在 prime subfield 上保持原來的對應，所以 $x^q - x$ 的係數保持不變，仍有
$$\widetilde f_i(x) \mid x^q - x.$$

$x^q - x$ 在 $F'$ 中完全分解，因此其正次數因子 $\widetilde f_i$ 在 $F'$ 中有一個根 $\beta_i$。  
利用剛才的同構延伸方法，得到
$$\phi_i : E_i = E_{i - 1}(\alpha_i)
\longrightarrow E'_{i - 1}(\beta_i) \subseteq F'.$$
它延伸 $\phi_{i - 1}$，並將 $\alpha_i$ 對應到 $\beta_i$。

重複有限次，最後得到一個保留加法與乘法的一一映射 $\phi_q : F \longrightarrow F'$。  
兩邊都只有 $q$ 個元素，所以它也映射到 $F'$ 的全部元素，因而 $F \cong F'$。

結合存在性與唯一性，對每個質數 $p$ 與正整數 $m$，
$$\boxed{\text{存在一個含有 }p^m\text{ 個元素的域，且在同構意義下唯一。}}$$
因此記號 $\mathbb F_{p^m}$ 表示這個同構類；選擇不同的建構，得到的運算可以透過同構互相對應。

分裂域與有限域分類的延伸閱讀：  
[Kopparty, Introduction to finite fields（§1.5–1.7）](https://www.math.toronto.edu/swastik/courses/finitefields-F23/intro.pdf)。

### 例子：$\mathbb F_8$ 的兩種多項式表示

考慮兩個域
$$A = \mathbb F_2[x]/(x^3 + x + 1),\qquad
B = \mathbb F_2[t]/(t^3 + t^2 + 1).$$
兩個三次多項式在 $0, 1$ 都不為零，因此都不可約。  
令 $\alpha \in A$ 與 $\beta \in B$ 分別是 $x$ 與 $t$ 的餘數類。

因為 $\alpha^3 + \alpha + 1 = 0$，所以
$$(\alpha^{-1})^3 + (\alpha^{-1})^2 + 1
= \alpha^{-3} (1 + \alpha + \alpha^3) = 0.$$

因此 $\beta$ 可以對應到 $A$ 中的 $\alpha^{-1}$。  
同構延伸公式具體寫成
$$\Phi : B \longrightarrow A,\qquad
a_0 + a_1\beta + a_2\beta^2
\longmapsto a_0 + a_1\alpha^{-1} + a_2\alpha^{-2},\qquad a_i \in \mathbb F_2.$$
這個對應保留兩種域表示的加法與乘法。  
兩個域各有 $8$ 個元素，這正是有限域唯一性的一個具體例子。

## 第一個 Extension Field：$\mathbb F_4$

### 把 $x$ 的餘數類命名為 $\alpha$

取不可約多項式 $f(x) = x^2 + x + 1$，定義
$$\mathbb F_4 = \mathbb F_2[x]/(x^2 + x + 1).$$
把 $x$ 的餘數類記為 $\alpha$，則
$$\alpha^2 + \alpha + 1 = 0,\qquad \alpha^2 = \alpha + 1.$$
每個元素都唯一寫成 $a + b\alpha$，其中 $a, b \in \mathbb F_2$，所以
$$\mathbb F_4 = \{0, 1, \alpha, \alpha + 1\}.$$
基底是 $1, \alpha$，而 characteristic 仍是 $2$。

### 加法：係數分別相加

例如
$$\alpha + (\alpha + 1) = 1,\qquad (\alpha + 1) + (\alpha + 1) = 0.$$
以 $0, 1, \alpha, \alpha + 1$ 的順序排列，加法表如下：

| $+$ | $0$ | $1$ | $\alpha$ | $\alpha + 1$ |
|---|---|---|---|---|
| $0$ | $0$ | $1$ | $\alpha$ | $\alpha + 1$ |
| $1$ | $1$ | $0$ | $\alpha + 1$ | $\alpha$ |
| $\alpha$ | $\alpha$ | $\alpha + 1$ | $0$ | $1$ |
| $\alpha + 1$ | $\alpha + 1$ | $\alpha$ | $1$ | $0$ |

### 乘法：用 $\alpha^2 = \alpha + 1$ 降低次數

乘法先展開，再使用 $\alpha^2 = \alpha + 1$。  
例如
$$\alpha(\alpha + 1) = \alpha^2 + \alpha = 1,$$
$$(\alpha + 1)^2 = \alpha^2 + 2\alpha + 1 = \alpha.$$
最後一式中的 $2\alpha = 0$，因為 characteristic 是 $2$。

三個非零元素的反元素為
$$1^{-1} = 1,\qquad \alpha^{-1} = \alpha + 1,\qquad (\alpha + 1)^{-1} = \alpha.$$

完整的乘法表如下；每個 entry 都是所在 row 與 column 的標記相乘。

| $\cdot$ | $0$ | $1$ | $\alpha$ | $\alpha + 1$ |
|---|---|---|---|---|
| $0$ | $0$ | $0$ | $0$ | $0$ |
| $1$ | $0$ | $1$ | $\alpha$ | $\alpha + 1$ |
| $\alpha$ | $0$ | $\alpha$ | $\alpha + 1$ | $1$ |
| $\alpha + 1$ | $0$ | $\alpha + 1$ | $1$ | $\alpha$ |

也可以把 $\alpha + 1$ 寫成 $\alpha^2$；此時 $\alpha^3 = 1$。

### 四個元素，兩種代數結構

比較 $\mathbb F_4$ 與 $\mathbb Z/4\mathbb Z$：

| 性質 | $\mathbb F_4$ | $\mathbb Z/4\mathbb Z$ |
|---|---|---|
| 元素個數 | $4$ | $4$ |
| $1 + 1$ | $0$ | $2$ |
| 非零元素的乘法反元素 | 每個都有 | $2$ 沒有 |
| 是否為域 | 是 | 否 |

運算規則是代數結構的一部分；元素個數相同的集合可以有不同的運算規則。

## 更多例子：$\mathbb F_8$ 與 $\mathbb F_9$

### 用三次多項式建構 $\mathbb F_8$

取 $f(x) = x^3 + x + 1 \in \mathbb F_2[x]$。  
因為 $f(0) = f(1) = 1$，它沒有根；次數為 $3$，所以不可約。

定義 $\mathbb F_8 = \mathbb F_2[x]/(x^3 + x + 1)$，把 $x$ 的餘數類記為 $\alpha$。  
於是
$$\alpha^3 = \alpha + 1.$$
每個元素唯一寫成 $a_0 + a_1\alpha + a_2\alpha^2$，其中 $a_i \in \mathbb F_2$。  
本節的 $\alpha$ 表示這個 $\mathbb F_8$ 中的元素。

利用 $\alpha^3 = \alpha + 1$，可以依序算出：

| 冪次 | 在基底 $1, \alpha, \alpha^2$ 下的表示 |
|---|---|
| $\alpha^0$ | $1$ |
| $\alpha^1$ | $\alpha$ |
| $\alpha^2$ | $\alpha^2$ |
| $\alpha^3$ | $\alpha + 1$ |
| $\alpha^4$ | $\alpha^2 + \alpha$ |
| $\alpha^5$ | $\alpha^2 + \alpha + 1$ |
| $\alpha^6$ | $\alpha^2 + 1$ |

再乘一次 $\alpha$，得到
$$\alpha^7 = \alpha^3 + \alpha = 1.$$
前面七個冪次恰好列出所有非零元素，因此乘法可以用指數計算：
$$\alpha^i\alpha^j = \alpha^{i + j},\qquad \text{指數取模 }7.$$
例如
$$\alpha^5\alpha^4 = \alpha^2,\qquad (\alpha^3)^{-1} = \alpha^4.$$
加法則用基底表示，把三個係數分別做 XOR。

### 手算一個乘法與一個除法

乘法可以直接展開：
$$(\alpha + 1)(\alpha^2 + 1)
= \alpha^3 + \alpha^2 + \alpha + 1
= \alpha^2.$$
也可以查冪次表：$\alpha + 1 = \alpha^3$、$\alpha^2 + 1 = \alpha^6$，因此乘積為 $\alpha^9 = \alpha^2$。

對除法，同樣使用反元素：
$$\frac{\alpha^2 + 1}{\alpha + 1}
= \frac{\alpha^6}{\alpha^3}
= \alpha^3 = \alpha + 1.$$

### 用 Python 表示與計算 $\mathbb F_8$

用整數 $a_0 + 2a_1 + 4a_2$ 標記元素 $a_0 + a_1\alpha + a_2\alpha^2$。  
例如 $\alpha$ 的標記是 $2$，$\alpha^2 + 1$ 的標記是 $5$。

這三個 binary digits 儲存多項式的係數，所以加法用 `a ^ b`。  
乘法則用下面的 shift-and-add；每次出現 $\alpha^3$，就用 $\alpha^3 + \alpha + 1 = 0$ 化簡。  
`0b1011` 表示多項式 $x^3 + x + 1$，輸入與輸出的標記都在 $0, \ldots, 7$ 中。

In [1]:
def gf8_mul(a, b):
    result = 0
    while b:
        if b & 1:
            result ^= a
        b >>= 1
        a <<= 1
        if a & 0b1000:
            a ^= 0b1011
    return result

In [2]:
powers = [1]
for _ in range(6):
    powers.append(gf8_mul(powers[-1], 2))
for k, value in enumerate(powers):
    print(f"alpha^{k}: {value:03b}")
print("alpha^7:", f"{gf8_mul(powers[-1], 2):03b}")

alpha^0: 001
alpha^1: 010
alpha^2: 100
alpha^3: 011
alpha^4: 110
alpha^5: 111
alpha^6: 101
alpha^7: 001


輸出的三個 digits 從左到右表示 $a_2, a_1, a_0$。  
例如 `110` 表示 $\alpha^2 + \alpha$，可以和前面的冪次表逐項對照。

下面計算剛才的乘法，再求 $\alpha + 1$ 的反元素。  
尋找反元素時，只要在七個非零元素中找出乘積為 $1$ 的那一個。

In [3]:
a, b = 0b011, 0b101
inverse_a = next(c for c in range(1, 8) if gf8_mul(a, c) == 1)
print("(alpha + 1)(alpha^2 + 1):", f"{gf8_mul(a, b):03b}")
print("inverse of alpha + 1:", f"{inverse_a:03b}")
print("(alpha^2 + 1)/(alpha + 1):", f"{gf8_mul(b, inverse_a):03b}")

(alpha + 1)(alpha^2 + 1): 100
inverse of alpha + 1: 110
(alpha^2 + 1)/(alpha + 1): 011


### Characteristic 為 $3$ 的例子：$\mathbb F_9$

在 $\mathbb F_3[x]$ 中，$x^2 + 1$ 在 $0, 1, 2$ 都不為零，所以不可約。  
令
$$\mathbb F_9 = \mathbb F_3[x]/(x^2 + 1),\qquad \beta^2 = -1 = 2.$$
九個元素是 $a + b\beta$，其中 $a, b \in \mathbb F_3$。  
加法與乘法的係數都取模 $3$，例如
$$(1 + \beta) + (2 + \beta) = 2\beta,$$
$$(1 + \beta)^2 = 1 + 2\beta + \beta^2 = 2\beta.$$

這個域的 characteristic 是 $3$，所以 $1 + 1 = 2$，$1 + 1 + 1 = 0$。

計算反元素時，可以先配出純量：
$$(1 + \beta)(1 - \beta) = 1 - \beta^2 = 2.$$
因為 $2^{-1} = 2$ 在 $\mathbb F_3$ 中成立，所以
$$(1 + \beta)^{-1} = 2(1 - \beta) = 2 + \beta.$$

## Multiplicative Groups 與 Primitive Elements

### 非零元素的乘法群

一個 **group（群）** 是具有結合律、單位元素與反元素的運算系統，且運算封閉。  
域的非零元素在乘法下形成交換群，記為
$$\mathbb F_q^\times = \mathbb F_q \setminus \{0\}.$$
它有 $q - 1$ 個元素。

對 $a \ne 0$，使 $a^r = 1$ 的最小正整數 $r$ 稱為 $a$ 的 **multiplicative order**。  
由群的 Lagrange theorem，$r$ 整除 $q - 1$，因此
$$a^{q - 1} = 1.$$

### Primitive element

**定理：** 有限域的乘法群是 cyclic。  
也就是存在一個非零元素 $\gamma$，使
$$\mathbb F_q^\times = \{1, \gamma, \gamma^2, \ldots, \gamma^{q - 2}\}.$$
這樣的 $\gamma$ 稱為 **primitive element**；它的 order 是 $q - 1$。

例如前面 $\mathbb F_4$ 的 $\alpha$ 有 order $3$，$\mathbb F_8$ 的 $\alpha$ 有 order $7$。  
兩者都能用冪次列出整個乘法群，前面的表已直接驗證它們是 primitive elements。  
一般有限域的 cyclicity 在這裡作為定理使用。

在 $\mathbb F_5$ 中，$2$ 的冪次為
$$2^0 = 1,\quad 2^1 = 2,\quad 2^2 = 4,\quad 2^3 = 3,\quad 2^4 = 1.$$
所以 $2$ 是 primitive element。  
元素 $4$ 滿足 $4^2 = 1$，order 只有 $2$。

選定 primitive element 後，非零元素的乘法與除法變成指數的加減：
$$\gamma^i\gamma^j = \gamma^{i + j},\qquad
\frac{\gamma^i}{\gamma^j} = \gamma^{i - j},\qquad \text{指數取模 }q - 1.$$
加法則需要回到係數表示，或另外使用加法表。

### 如何檢查 primitive element？

令 $N = q - 1$。  
非零元素 $\gamma$ 是 primitive element，當且僅當對 $N$ 的每個質因數 $\ell$ 都有
$$\gamma^{N/\ell} \ne 1.$$
因為若 order $r$ 是 $N$ 的真因數，就能找到某個質因數 $\ell$，使 $r$ 整除 $N/\ell$。

例如在 $\mathbb F_9$ 中，只需檢查 $\gamma^4 \ne 1$。  
取 $\gamma = 1 + \beta$，前面的計算給出
$$\gamma^2 = 2\beta,\qquad \gamma^4 = 2,\qquad \gamma^8 = 1.$$
因此 $1 + \beta$ 是 primitive element。

### Irreducible 與 primitive polynomials

若 $f(x)$ 是 monic irreducible polynomial，且 $x$ 的餘數類是所建構域的 primitive element，  
就稱 $f(x)$ 為 **primitive polynomial**。

例如 $x^2 + x + 1$ 與 $x^3 + x + 1$ 在 $\mathbb F_2$ 上都是 primitive polynomials。

在 $\mathbb F_3$ 上，$x^2 + 1$ 是 irreducible，但其根 $\beta$ 滿足
$$\beta^2 = 2,\qquad \beta^4 = 1.$$
所以 $\beta$ 的 order 是 $4$，小於 $|\mathbb F_9^\times| = 8$。  
不可約性保證建構出域；primitive 的條件還要求這個根能生成整個乘法群。

## Frobenius Map

### Characteristic $p$ 下的 $p$ 次方

在 characteristic 為 $p$ 的域中，二項式定理的中間係數 $\binom pk$ 都被 $p$ 整除，因此
$$(a + b)^p = a^p + b^p,\qquad (ab)^p = a^pb^p.$$
映射 $\varphi(a) = a^p$ 稱為 **Frobenius map**。  
它保留加法、乘法與 $1$，而且是一一對應。

一一性來自 $a^p = b^p \Longrightarrow (a - b)^p = 0 \Longrightarrow a = b$；  
在有限集合中，一一性也保證映射到全部元素。  
因此 Frobenius 是有限域的 automorphism。

### 每個元素都滿足 $a^q = a$

令 $q = p^m$。非零元素滿足 $a^{q - 1} = 1$，而零元素也滿足下式，因此
$$\boxed{a^q = a\qquad \text{對每個 }a \in \mathbb F_q.}$$
這表示連續做 $m$ 次 Frobenius map，會回到原來的元素：
$$\varphi^m(a) = a^{p^m} = a.$$
例如在前面的 $\mathbb F_8$ 中，平方會形成循環
$$\alpha\ \longmapsto\ \alpha^2\ \longmapsto\ \alpha^4\ \longmapsto\ \alpha^8 = \alpha.$$
這三個元素都是 $x^3 + x + 1$ 的根；係數在 $\mathbb F_2$ 中，所以 $f(a^2) = f(a)^2$。

## 在有限域上編碼

### 一個 $[3, 1, 3]_4$ linear code

回到 $\mathbb F_4 = \{0, 1, \alpha, \alpha^2\}$，其中 $\alpha^2 = \alpha + 1$、$\alpha^3 = 1$。  
沿用 row-vector 慣例，選擇 generator matrix
$$G = \begin{pmatrix}1 & \alpha & \alpha^2\end{pmatrix}.$$
資訊 symbol $u \in \mathbb F_4$ 編碼成
$$c = uG = (u, u\alpha, u\alpha^2).$$
因為 $G$ 的 rank 是 $1$，這個 code 有 $4^1 = 4$ 個 codewords。

四個 codewords 如下：

| $u$ | $uG$ |
|---|---|
| $0$ | $(0, 0, 0)$ |
| $1$ | $(1, \alpha, \alpha^2)$ |
| $\alpha$ | $(\alpha, \alpha^2, 1)$ |
| $\alpha^2$ | $(\alpha^2, 1, \alpha)$ |

所有非零 codewords 的 Hamming weight 都是 $3$，所以 minimum distance 是 $3$。  
這裡的 weight 與 distance 都數 symbols，參數是 $[3, 1, 3]_4$。

### Parity-check matrix 與 syndrome

在本課的慣例中，$H$ 是 $n \times (n - k)$ 的矩陣。  
對剛才的 code，可以取
$$H = \begin{pmatrix}
\alpha & \alpha^2 \\
1 & 0 \\
0 & 1
\end{pmatrix}.$$
直接計算得到 $GH = 0$；$H$ 的 rank 是 $2$，因此 $\{c : cH = 0\}$ 恰好是這個一維 code。

若傳送 $c$，收到 $y = c + e$，syndrome 為
$$s = yH = (c + e)H = eH.$$
若只有第 $j$ 個 symbol 錯誤，錯誤值是 $\varepsilon \ne 0$，則
$$s = \varepsilon h_j,$$
其中 $h_j$ 是 $H$ 的第 $j$ 個 row。  
所以 syndrome 包含錯誤位置與錯誤值的資訊。  
這個例子的三個 rows 互不成比例，能區分三個位置上的任意單一 symbol error。

### 解一個單一 symbol error

傳送 $c = (1, \alpha, \alpha^2)$，假設第二個位置加上錯誤值 $\alpha^2$。  
因為 $\alpha + \alpha^2 = 1$，收到
$$y = (1, 1, \alpha^2).$$
計算 syndrome：
$$s = yH = (\alpha + 1,\ \alpha^2 + \alpha^2) = (\alpha^2, 0).$$
它等於 $\alpha^2$ 乘上 $H$ 的第二個 row，因此得知位置是 $2$、錯誤值是 $\alpha^2$。  
最後減去 error vector，得到
$$y - (0, \alpha^2, 0) = (1, \alpha, \alpha^2) = c.$$

### 預告：用取值建構 Reed–Solomon codes

在 $\mathbb F_q$ 中選 $n$ 個不同的點 $a_1, \ldots, a_n$，令 $1 \le k \le n \le q$。  
把次數小於 $k$ 的多項式編碼成
$$f(x)\ \longmapsto\ (f(a_1), \ldots, f(a_n)).$$
兩個不同多項式之差的次數至多為 $k - 1$，因此至多在 $k - 1$ 個位置取相同的值。  
所以 minimum distance 至少是 $n - k + 1$，且取值映射是一一的，code 的 dimension 是 $k$。

取 $f(x) = \prod_{i = 1}^{k - 1}(x - a_i)$，它在恰好 $k - 1$ 個選定位置為零，所以
$$\boxed{[n, k, n - k + 1]_q.}$$
這就是 Reed–Solomon code 的基本建構；$k = 1$ 時，上面的空乘積取為 $1$。

例如在 $\mathbb F_4$ 上，選取值點 $0, 1, \alpha, \alpha^2$，並令 $f(x) = u + vx$。  
得到 generator matrix
$$G = \begin{pmatrix}
1 & 1 & 1 & 1 \\
0 & 1 & \alpha & \alpha^2
\end{pmatrix}.$$
每個非零一次多項式至多有一個根，且 $f(x) = x$ 的取值向量恰好有三個非零 symbols。  
所以這個 code 的參數是 $[4, 2, 3]_4$，有 $4^2 = 16$ 個 codewords。  
較大的有限域讓我們能選擇更多不同的取值點。

## Cyclic Codes

### 循環位移與 cyclic code

對長度為 $n$ 的向量，定義向右的 **cyclic shift（循環位移）**：
$$\sigma(c_0, c_1, \ldots, c_{n - 1})
= (c_{n - 1}, c_0, \ldots, c_{n - 2}).$$
若 linear code $\mathcal C \subseteq \mathbb F_q^n$ 對這個位移封閉，就稱為 **cyclic code**：
$$c \in \mathcal C\quad\Longrightarrow\quad \sigma(c) \in \mathcal C.$$
本節將座標編號寫成 $0, 1, \ldots, n - 1$，方便和多項式的冪次對應。

重複做 $n$ 次循環位移，就回到原來的向量，所以 $\sigma^n$ 是 identity。  
若 $\mathcal C$ 是 cyclic code，則 $\sigma^j(c) \in \mathcal C$ 對所有 $j$ 都成立。  
向左位移一次等於向右位移 $n - 1$ 次，因此也保持 code。

Cyclicity 是對指定的座標順序而言。  
若重新排列座標，循環位移所對應的 permutation 也會改變。

### Repetition 與 single parity-check codes

Repetition code 的每個 codeword 都具有形式
$$c = (a, a, \ldots, a),\qquad a \in \mathbb F_q.$$
循環位移後仍是同一個向量，所以 repetition code 是 cyclic。

Single parity-check code 定義為
$$\mathcal C = \left\{c \in \mathbb F_q^n : \sum_{i = 0}^{n - 1}c_i = 0\right\}.$$
循環位移只是重新排列相加的順序，總和仍為零，所以這個 code 也是 cyclic。

### 用多項式表示 cyclic shift

把 codeword 的 symbols 當作多項式的係數：
$$c = (c_0, \ldots, c_{n - 1})
\quad\longleftrightarrow\quad
c(z) = c_0 + c_1z + \cdots + c_{n - 1}z^{n - 1}.$$
每個 $z$ 的冪次記錄一個座標位置。  
這些多項式可以看成 quotient ring $\mathbb F_q[z]/(z^n - 1)$ 中的元素。

乘上 $z$，再利用 $z^n = 1$ 取餘數，得到
$$zc(z) \equiv c_{n - 1} + c_0z + \cdots + c_{n - 2}z^{n - 1}
\pmod{z^n - 1}.$$
它的係數向量恰好是 $\sigma(c)$。

因此，cyclic code 的多項式表示對「乘上 $z$ 並取模 $z^n - 1$」封閉。  
再利用線性性，它對乘上任意多項式並取模也封閉。

### 用 primitive element 排列座標

取 $\mathbb F_{2^m}$ 的 primitive element $\alpha$，並令 $n = 2^m - 1$。  
這個域的全部非零元素可以依序排列成
$$1, \alpha, \alpha^2, \ldots, \alpha^{n - 1},\qquad \alpha^n = 1.$$
如果用它們標記 $n$ 個座標，座標標記乘上 $\alpha$，就沿著這個順序移動一格。  
從最後的 $\alpha^{n - 1}$ 乘上 $\alpha$，則回到第一個標記 $1$。

將這個域看成 $\mathbb F_2$ 上的 $m$ 維向量空間，就能把這些標記和非零 binary vectors 對應起來。  
這個座標順序會用在本次的 punctured RM code 作業中。

## 週五作業

### 第一題：Punctured RM code 的 cyclicity

令 $m \ge 2$、$0 \le r \le m$，並令 $n = 2^m - 1$。  
固定 $\mathbb F_{2^m}$ 在 $\mathbb F_2$ 上的一組基底 $b_1, \ldots, b_m$，定義對應
$$\iota : \mathbb F_2^m \longrightarrow \mathbb F_{2^m},\qquad
\iota(x_1, \ldots, x_m) = \sum_{j = 1}^m x_j b_j.$$
取 $\mathbb F_{2^m}$ 的 primitive element $\alpha$。

將 $\mathrm{RM}(r, m)$ 在零向量 $x = 0$ 的座標刪除，  
再按標記 $1, \alpha, \ldots, \alpha^{n - 1}$ 的順序排列剩下的座標。

得到長度為 $n$ 的 **punctured RM code**：
$$\mathrm{RM}(r, m)^*
= \left\{\left(f(\iota^{-1}(\alpha^i))\right)_{i = 0}^{n - 1}
: f \in \mathbb F_2[X_1, \ldots, X_m],\ \deg f \le r\right\}.$$
這裡 $f$ 在 binary vectors 上取值；可以只考慮 multilinear polynomials。

**證明 $\mathrm{RM}(r, m)^*$ 在這個座標順序下是 binary cyclic code。**

提示：考慮映射 $a \mapsto \alpha a$ 在 $\mathbb F_2$ 上的線性性，  
並使用 0921 作業中可逆線性變換保持 $\mathrm{RM}(r, m)$ 的結論。

### 週一作業